crates/apollo-parser/src/parser/grammar/document.rs

```diff
@@ -13,16 +14,19 @@ use crate::{
 pub(crate) fn document(p: &mut Parser) {
     let doc = p.start_node(SyntaxKind::DOCUMENT);
 
-    while let Some(node) = p.peek() {
+    p.peek_while(|p, kind| {
         assert_eq!(
             p.recursion_limit.current, 0,
             "unbalanced limit increment / decrement"
         );
 
-        match node {
+        match kind {
             TokenKind::StringValue => {
-                let def = p.peek_data_n(2).unwrap();
-                select_definition(def, p);
+                if let Some(def) = p.peek_data_n(2) {
```

Comment:
Could return `None` if the token limit was reached

Reply:
Then you would get a `Some(Eof)` here. that's something i forgot to mention in the PR description: a token limit immediately cuts off the stream, while end of file is its own token. I considered emitting an Eof token when a limit is reached, and I do think that could make sense, but I felt that removing the unwraps from the parser was worthwhile anyways--`peek_data_n(2)` would not panic if there was always an Eof token still in the queue, but relying on that detail implicitly throughout the codebase is not great. So in this PR I did not add an Eof token when the limit is reached.